src/MizarData/FrameTrackManagerTest.cpp

```diff
@@ -0,0 +1,233 @@
+                                                              PresentEvent::kD3d9};
+
+const absl::flat_hash_map<PresentEvent::Source, std::vector<FrameStartNs>> kEtwSourceToFrameStart =
+    MakeMap(kEtwSources,
+            std::vector<std::vector<FrameStartNs>>{kDxgiFrameStarts, kD3d9FrameStarts});
+const absl::flat_hash_map<PresentEvent::Source, std::vector<PresentEvent>>
+    kEtwSourceToPresentEvent = MakeMap(
+        kEtwSources, std::vector<std::vector<PresentEvent>>{MakePresentEvent(kDxgiFrameStarts),
+                                                            MakePresentEvent(kD3d9FrameStarts)});
+
+static std::pair<std::vector<ScopeInfo>, std::vector<PresentEvent::Source>> DecomposeSources(
+    const absl::flat_hash_map<FrameTrackId, FrameTrackInfo>& id_to_infos) {
+  std::vector<ScopeInfo> scope_id_infos;
+  std::vector<PresentEvent::Source> etw_sources;
+
+  for (const auto& [unused_id, info] : id_to_infos) {
+    Visit([&scope_id_infos](const ScopeInfo& info) { scope_id_infos.push_back(info); },
+          [&etw_sources](PresentEvent::Source source) { etw_sources.push_back(source); }, info);
+  }
+
+  return {scope_id_infos, etw_sources};
+}
+
+class ExpectFrameTracksHasFrameStartsForScopes : public ::testing::Test {
+ public:
+  ExpectFrameTracksHasFrameStartsForScopes() {
+    EXPECT_CALL(data_, GetCaptureData).WillRepeatedly(ReturnRef(capture_data_));
+  }
+
+  void SourceToPresentEventExpectCall() {
+    EXPECT_CALL(data_, source_to_present_events).WillRepeatedly(Return(kEtwSourceToPresentEvent));
+  }
+
+  void CaptureDataExpectCalls() {
+    EXPECT_CALL(capture_data_, GetAllProvidedScopeIds).WillRepeatedly(Return(kScopeIds));
+    EXPECT_CALL(capture_data_, GetScopeInfo).WillRepeatedly(Invoke([](const ScopeId id) {
+      return kScopeIdToInfo.at(id);
+    }));
+    EXPECT_CALL(capture_data_, GetTimersForScope)
+        .WillRepeatedly(Invoke([](const ScopeId id, uint64_t /*min*/, uint64_t /*max*/) {
+          return kScopeIdToSTimerInfoPtrs.at(id);
+        }));
+  }
+
+  void ExpectGetFrameStartsIsCorrect(FrameStartNs min_start, FrameStartNs max_start) {
+    for (const auto& [id, info] : frame_track_manager_.GetFrameTracks()) {
+      const std::vector<FrameStartNs> expected = Visit(
+          [](const ScopeInfo& info) { return kScopeInfoToFrameStarts.at(info); },
+          [min_start, max_start](PresentEvent::Source source) {
+            std::vector<FrameStartNs> expected;
+            const std::vector<FrameStartNs> all_frame_starts = kEtwSourceToFrameStart.at(source);
+            std::copy_if(std::begin(all_frame_starts), std::end(all_frame_starts),
+                         std::back_inserter(expected), [min_start, max_start](FrameStartNs start) {
+                           return min_start <= start && start <= max_start;
+                         });
+            return expected;
+          },
+          info);
+
+      const std::vector<FrameStartNs> actual =
```

Comment:
again, please consider renaming this variable

Reply:
actual_start_time_list

Reply:
actual_start_time_list

Reply:
actual_frame_starts